### ISSUE TITLE
Tensorflow is pulling enum34 in Python 3.5 and 3.6, which breaks code completion in Spyder

### ISSUE THREAD

**Issue by ccordoba12** [tensorflow/tensorflow#15136]
### System information
- **Have I written custom code (as opposed to using a stock example script provided in TensorFlow)**: No
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Linux Ubuntu 17.04
- **TensorFlow installed from (source or binary)**: binary
- **TensorFlow version (use command below)**: 1.4.0
- **Python version**: 3.5, 3.6
- **Bazel version (if compiling from source)**: N/A
- **GCC/Compiler version (if compiling from source)**: N/A
- **CUDA/cuDNN version**: N/A
- **GPU model and memory**: N/A
- **Exact command to reproduce**: `pip install tensorflow`

### Describe the problem
When you run

    pip install tensorflow

in a virtualenv with Python 3.5 or 3.6, one of the dependencies pulled by `pip` is `enum34`, as the following console output shows

```
$ pip install tensorflow
Collecting tensorflow
  Downloading tensorflow-1.4.0-cp35-cp35m-manylinux1_x86_64.whl (40.7MB)
    100% |████████████████████████████████| 40.7MB 39kB/s 
Collecting numpy>=1.12.1 (from tensorflow)
  Using cached numpy-1.13.3-cp35-cp35m-manylinux1_x86_64.whl
Collecting six>=1.10.0 (from tensorflow)
  Using cached six-1.11.0-py2.py3-none-any.whl
Collecting protobuf>=3.3.0 (from tensorflow)
  Downloading protobuf-3.5.0.post1-cp35-cp35m-manylinux1_x86_64.whl (6.4MB)
    100% |████████████████████████████████| 6.4MB 217kB/s 
Requirement already satisfied: wheel>=0.26 in ./.virtualenvs/tf/lib/python3.5/site-packages (from tensorflow)
Collecting enum34>=1.1.6 (from tensorflow)
  Downloading enum34-1.1.6-py3-none-any.whl
```

This package is only necessary if `Python < 3.4`, as described here:

https://pypi.python.org/pypi/enum34/1.1.6

so it's an error that `tensorflow` pullis it for Python 3.5 and 3.6.

Besides, this package breaks Spyder code completion machinery in its Editor, as it has been verified by several users. See for example:

spyder-ide/spyder#5782.

**Comment by drpngx** [comment]
@gunan @jart should we update the package?

**Comment by gunan** [comment]
@av8ramit is working on 1.4.1
If we have a fix, we can cherrypick this into 1.4.1
However, the only attempted fix ended up breaking the pip package.
if we have a fix that works for all python versions, I am happy to release it with 1.4.1

**Comment by gunan** [comment]
Also CC @yifeif

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by gunan** [comment]
I think our final decision on this was to remove the dependency on `enum34` library.
@martinwicke to confirm the decision
@yifeif @angersson to make sure it is removed for 1.5 release
@xiejw to FYI

**Comment by xiejw** [comment]
Thanks for the heads up. This sounds fine to me.

**Comment by martinwicke** [comment]
Confirmed.

**Comment by yifeif** [comment]
For 1.5, enum should only be included as a dependency if the python version is < 3.4 https://github.com/tensorflow/tensorflow/blob/r1.5/tensorflow/tools/pip_package/setup.py#L67

**Comment by martinwicke** [comment]
I think the idea is to remove all use of enum, no?

**Comment by yifeif** [comment]
Just checked with tf-nightly, enum34 is only installed with python2.7. Let me know if this is not what we wanted.

**Comment by gunan** [comment]
Do we know if it is easy to remove all of our dependency on enum?

**Comment by martinwicke** [comment]
I think there's only a two uses. One in variable_scope.py, as superclass of a private class _ReuseMode. The other in xla_client.py, as superclass of PaddingType, which is not part of the public API. 

So it should be easy to get rid of entirely.

**Comment by tensorflowbutler** [comment]
Nagging Assigneee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
A member of the TensorFlow organization has replied after the stat:awaiting tensorflower label was applied.

**Comment by drpngx** [comment]
@ccordoba12 would you be willing to submit a PR?

**Comment by ccordoba12** [comment]
Sure, could you point me to the places I have to look for?

**Comment by deepankar27** [comment]
@tensorflowbutler Waiting for the build w/o enum

**Comment by luchensk** [comment]
Waiting +1

**Comment by gunan** [comment]
I think the enum dependency should not be fetched anymore for later python versions, @yifeif to confirm.
But we should still kill the dependency on enum34.

**Comment by yifeif** [comment]
We shouldn't be depending on enum34 for python 3 anymore. Let me know if that is not the case? Should we close this issue or keep it open for removing enum34 entirely at some point @gunan?

**Comment by gunan** [comment]
Let's keep it open to remove enum34 entirely.

**Comment by tensorflowbutler** [comment]
Please remove the assignee, as this issue is inviting external contributions. Otherwise, remove the `contributions welcome` label. Thank you.

**Comment by smartb-pair** [comment]
This is causing a lot of trouble for us on Python 3.6.3 - there are workarounds but they are...unpleasant.

**Comment by yifeif** [comment]
@smartb-pair what kind of issue are you seeing? python3.6 shouldn't be pulling enum as a dependency anymore.

**Comment by smartb-pair** [comment]
The dep on enum is declared in the metadata for the tensorflow wheel file provided by Google (we need to install from this file rather then pypi for a variety of org-specific reasons).

Blake

**Comment by ccordoba12** [comment]
@smartb-pair, are you installing old Tensorflow versions? I think this was fixed in version 1.5.

**Comment by smartb-pair** [comment]
We currently have a requirement for 1.4.1 - will see if we can rev up...

**Comment by wedesoft** [comment]
Not sure whether it will help others but I managed to install Tensorflow as follows:
```
sudo pip install --ignore-installed tensorflow
```

**Comment by atharvanvaidya** [comment]
@wedesoft Well **THAT** worked for me! Thanks a Lot!!

**Comment by wt-huang** [comment]
Closing as this is resolved

**Comment by ccordoba12** [comment]
Thanks!

**Comment by pwais** [comment]
@wt-huang  nope!  2019 and this is still a bug:
```
$ docker run -it --rm tensorflow/tensorflow:1.15.0-py3-jupyter bash

...

root@2d0b0b6ff8ac:/tf# pip uninstall -y enum34
Uninstalling enum34-1.1.6:
  Successfully uninstalled enum34-1.1.6
```

**Comment by pwais** [comment]
cc https://github.com/tensorflow/tensorflow/issues/18480
